vscode-trace-webviews/src/trace-explorer/opened-traces/vscode-trace-explorer-opened-traces-widget.tsx

```diff
@@ -57,20 +59,24 @@ class TraceExplorerOpenedTraces extends React.Component<{}, OpenedTracesAppState
       };
       this._signalHandler = new VsCodeMessageManager();
       window.addEventListener('message', event => {
-
           const message = event.data; // The JSON data our extension sent
           switch (message.command) {
           case VSCODE_MESSAGES.SET_TSP_CLIENT:
-              const tspClientProvider: ITspClientProvider = new TspClientProvider(message.data, this._signalHandler);
+              this._urlProvider = new TraceServerUrlProvider();
+              const tspClientProvider: ITspClientProvider = new TspClientProvider(message.data, this._signalHandler, this._urlProvider);
               this._experimentManager = tspClientProvider.getExperimentManager();
-              this.setState({ tspClientProvider: tspClientProvider });
-              if (this.state.tspClientProvider) {
-                  this.state.tspClientProvider.addTspClientChangeListener(() => {
-                      if (this.state.tspClientProvider) {
-                          this._experimentManager = this.state.tspClientProvider.getExperimentManager();
-                      }
-                  });
-              }
+
+              const callback = () => {
+                  if (this.state.tspClientProvider) {
+                      this.state.tspClientProvider.addTspClientChangeListener(() => {
+                          if (this.state.tspClientProvider) {
+                              this._experimentManager = this.state.tspClientProvider.getExperimentManager();
+                          }
+                      });
+                  }
+              };
+
+              this.setState({ tspClientProvider: tspClientProvider }, callback);
```

Comment:
This looks like it could be a memory leak. What happens when a new `TspClientProvider` is created? The old one will still have the callback that was added to it, and that could cause it to be retained even though we're done with it.

The same is perhaps true of other cases where new `TspClientProviders` are created, but it's less clear who's subscribing to their events.

Reply:
Please correct me if I am wrong. If you are talking about the callback at line 69, then it will only be run when a new tspClientProvider is set. If you are talking about the listener on line 71, I was just trying to keep things the same as the existing code here and in the `theia-trace-extension`. Perhaps it is an issue that is outside of the scope for this PR. If you think so, please let me know and I can create an issue in GitHub for that.

Reply:
The way I understand the code, when a new TspClientProvider is created, a listener is added to itself, so no one else has a reference to it. When the old TspClientProvider is replaced and there are no more references to it, it should be discarded along with its listener.

The usual error pattern for such leaks is when the listener is added to an external object that remains when the object that added the listener is disposed...

It appears though that if a zombie TspClientProvider receives a client change event, its listener will execute, but it will set the experiment manager to the 'new' TspClientProvider's experiment manager (in addition to that new TspClientProvider also doing the same thing). So it could double set to the same value, maybe harmless? To avoid that there could be a check that the current state's tspClientProvider is the same one that created that particular listener. But I'm not sure if it's possible for the non-current TspClientProvider to receive a client change event.

Reply:
It seem unlikely that it will result in a memory leak in practice.